playwright/src/main/java/com/microsoft/playwright/impl/junit/BrowserContextExtension.java

```diff
@@ -66,10 +64,135 @@ public static BrowserContext getOrCreateBrowserContext(ExtensionContext extensio
   }
 
+  @Override
+  public void testSuccessful(ExtensionContext extensionContext) {
+    saveTraceWhenOn(extensionContext);
+    saveScreenshotWhenOn(extensionContext);
+    cleanupResources();
+  }
+
+  @Override
+  public void testAborted(ExtensionContext extensionContext, Throwable cause) {
+    saveTraceWhenOn(extensionContext);
+    saveScreenshotWhenOn(extensionContext);
+    cleanupResources();
+  }
+
+  @Override
+  public void testFailed(ExtensionContext extensionContext, Throwable cause) {
+    Options options = OptionsExtension.getOptions(extensionContext);
+    if (options.trace.equals(Options.Trace.ON) || options.trace.equals(Options.Trace.RETAIN_ON_FAILURE)) {
+      saveTrace(extensionContext);
+    }
+
+    if (options.screenshot.equals(Options.Screenshot.ON) || options.screenshot.equals(Options.Screenshot.ONLY_ON_FAILURE)) {
+      saveScreenshot(extensionContext, true);
+    }
+    cleanupResources();
+  }
+
+  private void cleanupResources() {
+    closePage();
+    closeBrowserContext();
+  }
+
+  private static void saveScreenshotWhenOn(ExtensionContext extensionContext) {
+    Options options = OptionsExtension.getOptions(extensionContext);
+    if (options.screenshot.equals(Options.Screenshot.ON)) {
+      saveScreenshot(extensionContext, false);
+    }
+  }
+
+  private static void saveTraceWhenOn(ExtensionContext extensionContext) {
+    Options options = OptionsExtension.getOptions(extensionContext);
+    if (options.trace.equals(Options.Trace.ON)) {
+      saveTrace(extensionContext);
+    }
+  }
+
+  private static void saveScreenshot(ExtensionContext extensionContext, boolean didTestFail) {
+    BrowserContext browserContext = threadLocalBrowserContext.get();
+    if (browserContext == null) {
+      return;
+    }
+    String fileNamePrefix = "test-finished-";
+
+    if (didTestFail) {
+      fileNamePrefix = "test-failed-";
+    }
+
+    Path outputPath = getOutputPath(extensionContext);
+    createOutputPath(outputPath);
+    for (int i = 0; i < browserContext.pages().size(); i++) {
+      Path screenshotPath = outputPath.resolve(fileNamePrefix + (i + 1) + ".png");
+      browserContext.pages().get(i).screenshot(new Page.ScreenshotOptions().setPath(screenshotPath));
+    }
+
+  }
+
+  private static void saveTrace(ExtensionContext extensionContext) {
+    BrowserContext browserContext = threadLocalBrowserContext.get();
+    if (browserContext == null) {
+      return;
+    }
+    Path outputPath = getOutputPath(extensionContext);
+    createOutputPath(outputPath);
+    Tracing.StopOptions stopOptions = new Tracing.StopOptions().setPath(outputPath.resolve("trace.zip"));
+    browserContext.tracing().stop(stopOptions);
+  }
+
+  private static void createOutputPath(Path outputPath) {
+    try {
+      Files.walk(outputPath).sorted(Comparator.reverseOrder()).map(Path::toFile).forEach(File::delete);
+      Files.createDirectories(outputPath);
+    } catch (NoSuchFileException ignored) {
+      // swallow
```

Comment:
This doesn't seem right, maybe wrap deletion into try catch but move createDirectories outside?

Reply:
I was doing too much.  I think it's cleaned up now where if the directory exists, I don't touch it.